src/client/datascience/notebookStorage/baseModel.ts

```diff
@@ -28,6 +29,47 @@ type KernelIdListEntry = {
     kernelId: string | undefined;
 };
 
+
+@injectable()
+export class PreferredRemoteKernelIdProvider {
+    constructor(
+        @inject(IMemento) @named(GLOBAL_MEMENTO) private readonly globalMemento: Memento,
+        @inject(ICryptoUtils) private crypto: ICryptoUtils
+    ) {}
+
+    public getPreferredRemoteKernelId(uri: Uri): string | undefined {
+        // Stored as a list so we don't take up too much space
+        const list: KernelIdListEntry[] = this.globalMemento.get<KernelIdListEntry[]>(ActiveKernelIdList, []);
+        if (list) {
+            // Not using a map as we're only going to store the last 40 items.
+            const fileHash = this.crypto.createHash(uri.toString(), 'string');
+            const entry = list.find((l) => l.fileHash === fileHash);
+            return entry?.kernelId;
+        }
+    }
+
+    public async storePreferredRemoteKernelId(uri: Uri, id: string | undefined): Promise<void> {
+        const list: KernelIdListEntry[] = this.globalMemento.get<KernelIdListEntry[]>(ActiveKernelIdList, []);
+        const fileHash = this.crypto.createHash(uri.toString(), 'string');
+        const index = list.findIndex((l) => l.fileHash === fileHash);
+        // Always remove old spot (we'll push on the back for new ones)
+        if (index >= 0) {
+            list.splice(index, 1);
+        }
+
+        // If adding a new one, push
+        if (id) {
+            list.push({ fileHash, kernelId: id });
+        }
+
+        // Prune list if too big
+        while (list.length > MaximumKernelIdListSize) {
```

Comment:
40 may not be enough. We might have telemetry on the number notebooks in a workspace though.

Reply:
I'll bump it to 100, 100 in a list is super small

Reply:
Updated to 100 & added telemetry to capture the size of this list